File: advent2017/extras/cxx/src/days/day06.cpp
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <openssl/sha.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

namespace {

const std::string kExpectedSha = "489246369534515a9df814e8824f41c427d6c02ab31d7b5c07cbdc935497f2ba";
const std::string kExpectedPart1 = "12841";
const std::string kExpectedPart2 = "8038";

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    const std::vector<std::string> cands{"advent2017/Day6/d6_input.txt", "Day6/d6_input.txt", "../Day6/d6_input.txt", "../../Day6/d6_input.txt"};
    for (const auto& c : cands) if (std::filesystem::exists(c)) return c;
    throw std::runtime_error("input not found");
}

std::string read_all(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) throw std::runtime_error("failed to read input");
    return std::string((std::istreambuf_iterator<char>(in)), {});
}

std::string sha256_hex(const std::string& data) {
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(data.data()), data.size(), out);
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (auto b : out) oss << std::setw(2) << static_cast<int>(b);
    return oss.str();
}

std::vector<int> parse_banks(const std::string& raw) {
    std::istringstream in(raw);
    std::vector<int> out;
    int v = 0;
    while (in >> v) out.push_back(v);
    return out;
}

std::string key_of(const std::vector<int>& a) {
    std::ostringstream oss;
    for (std::size_t i = 0; i < a.size(); ++i) {
        if (i) oss << ',';
        oss << a[i];
    }
    return oss.str();
}

void redistribute(std::vector<int>& a) {
    std::size_t idx = 0;
    for (std::size_t i = 1; i < a.size(); ++i) {
        if (a[i] > a[idx]) idx = i;
    }
    int blocks = a[idx];
    a[idx] = 0;

    const int n = static_cast<int>(a.size());
    const int q = blocks / n;
    const int r = blocks % n;
    if (q > 0) {
        for (auto& v : a) v += q;
    }
    for (int k = 1; k <= r; ++k) {
        a[(idx + static_cast<std::size_t>(k)) % a.size()]++;
    }
}

long long solve_part1(const std::vector<int>& banks) {
    auto a = banks;
    std::unordered_set<std::string> seen;
    long long steps = 0;
    while (seen.find(key_of(a)) == seen.end()) {
        seen.insert(key_of(a));
        redistribute(a);
        steps++;
    }
    return steps;
}

long long solve_part2(const std::vector<int>& banks) {
    auto a = banks;
    std::unordered_map<std::string, long long> seen;
    long long steps = 0;
    while (seen.find(key_of(a)) == seen.end()) {
        seen[key_of(a)] = steps;
        redistribute(a);
        steps++;
    }
    return steps - seen[key_of(a)];
}

}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part") part = std::stoi(argv[++i]);
        else if (a == "--input") input = argv[++i];
        else throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");

    input = resolve_input(input);
    const std::string raw = read_all(input);
    if (sha256_hex(raw) != kExpectedSha) throw std::runtime_error("checksum mismatch");
    const auto banks = parse_banks(raw);

    auto t0 = std::chrono::steady_clock::now();
    const std::string ans = std::to_string((part == 1) ? solve_part1(banks) : solve_part2(banks));
    const std::string expected = (part == 1) ? kExpectedPart1 : kExpectedPart2;
    if (ans != expected) throw std::runtime_error("answer mismatch");

    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    const double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=6 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day06.rs
use sha2::{Digest, Sha256};
use std::collections::{HashMap, HashSet};
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "489246369534515a9df814e8824f41c427d6c02ab31d7b5c07cbdc935497f2ba";
const EXPECTED_PART1: &str = "12841";
const EXPECTED_PART2: &str = "8038";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() {
        return provided.to_string();
    }
    let cands = ["advent2017/Day6/d6_input.txt", "Day6/d6_input.txt", "../Day6/d6_input.txt", "../../Day6/d6_input.txt"];
    for c in cands {
        if Path::new(c).exists() {
            return c.to_string();
        }
    }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new();
    h.update(&b);
    hex::encode(h.finalize())
}

fn parse_banks(raw: &str) -> Vec<i32> {
    raw.split_whitespace().map(|x| x.parse::<i32>().expect("int")).collect::<Vec<i32>>()
}

fn redistribute(a: &mut [i32]) {
    let mut idx: usize = 0;
    for i in 1..a.len() {
        if a[i] > a[idx] {
            idx = i;
        }
    }
    let blocks = a[idx];
    a[idx] = 0;

    let n = a.len() as i32;
    let q = blocks / n;
    let r = blocks % n;
    if q > 0 {
        for v in a.iter_mut() {
            *v += q;
        }
    }
    for k in 1..=r {
        let j = ((idx as i32 + k) % n) as usize;
        a[j] += 1;
    }
}

fn solve_part1(banks: &[i32]) -> i64 {
    let mut a = banks.to_vec();
    let mut seen: HashSet<Vec<i32>> = HashSet::new();
    let mut steps: i64 = 0;
    while !seen.contains(&a) {
        seen.insert(a.clone());
        redistribute(&mut a);
        steps += 1;
    }
    steps
}

fn solve_part2(banks: &[i32]) -> i64 {
    let mut a = banks.to_vec();
    let mut seen: HashMap<Vec<i32>, i64> = HashMap::new();
    let mut steps: i64 = 0;
    while !seen.contains_key(&a) {
        seen.insert(a.clone(), steps);
        redistribute(&mut a);
        steps += 1;
    }
    steps - seen[&a]
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => {
                part = args[i + 1].parse().unwrap();
                i += 2;
            }
            "--input" => {
                input = args[i + 1].clone();
                i += 2;
            }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 {
        panic!("--part 1|2 required");
    }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA {
        panic!("checksum mismatch");
    }
    let banks = parse_banks(&fs::read_to_string(&in_path).expect("read_to_string"));

    let t0 = Instant::now();
    let ans = if part == 1 { solve_part1(&banks) } else { solve_part2(&banks) }.to_string();
    let expected = if part == 1 { EXPECTED_PART1 } else { EXPECTED_PART2 };
    if ans != expected {
        panic!("answer mismatch");
    }

    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=6 part={} runtime_ms={:.3}", part, ms);
}
